diff --git a/src/EditProfileDialog.ui b/src/EditProfileDialog.ui
index 82298cf..66a1084 100644
--- a/src/EditProfileDialog.ui
+++ b/src/EditProfileDialog.ui
@@ -1109,6 +1109,11 @@
                 <string comment="The shape of the cursor">Underline</string>
                </property>
               </item>
+              <item>
+               <property name="text">
+                <string comment="The shape of the cursor">Custom</string>
+               </property>
+              </item>
              </widget>
             </item>
             <item>
diff --git a/src/Enumeration.h b/src/Enumeration.h
index a9fab40..fe4f722 100644
--- a/src/Enumeration.h
+++ b/src/Enumeration.h
@@ -85,7 +85,8 @@ public:
          */
         IBeamCursor     = 1,
         /** Draw a line underneath the cursor's position. */
-        UnderlineCursor = 2
+        UnderlineCursor = 2,
+        CustomCursor = 3
     };
 
     /** This enum describes the behavior of triple click action . */
diff --git a/src/TerminalDisplay.cpp b/src/TerminalDisplay.cpp
index 2235047..e75a381 100644
--- a/src/TerminalDisplay.cpp
+++ b/src/TerminalDisplay.cpp
@@ -632,6 +632,7 @@ void TerminalDisplay::drawBackground(QPainter& painter, const QRect& rect, const
     painter.fillRect(scrollBarArea, _scrollBar->palette().background());
 }
 
+#include <iostream>
 void TerminalDisplay::drawCursor(QPainter& painter,
                                  const QRect& rect,
                                  const QColor& foregroundColor,
@@ -678,6 +679,25 @@ void TerminalDisplay::drawCursor(QPainter& painter,
                          cursorRect.top(),
                          cursorRect.left(),
                          cursorRect.bottom());
+
+    } else if (_cursorShape == Enum::CustomCursor) {
+
+    	int centre = cursorRect.left() + ((cursorRect.right() - cursorRect.left()) / 2);
+        painter.drawLine(centre,
+                         cursorRect.top(),
+                         centre,
+                         cursorRect.bottom());
+
+	centre = cursorRect.top() + ((cursorRect.bottom() - cursorRect.top()) / 3);
+    	std::cout << "Top = " << cursorRect.top() << "\n" << 
+    	"Bottom = " << cursorRect.bottom() << "\n" <<
+    	"Centre = " << centre << std::endl;
+	
+        painter.drawLine(cursorRect.left(),
+                         centre,
+                         cursorRect.right(),
+                         centre);
+
     }
 }
 
diff --git a/src/ViewManager.cpp b/src/ViewManager.cpp
index 0b43bb8..9a7997a 100644
--- a/src/ViewManager.cpp
+++ b/src/ViewManager.cpp
@@ -858,6 +858,8 @@ void ViewManager::applyProfileToView(TerminalDisplay* view , const Profile::Ptr
         view->setKeyboardCursorShape(Enum::IBeamCursor);
     else if (cursorShape == Enum::UnderlineCursor)
         view->setKeyboardCursorShape(Enum::UnderlineCursor);
+    else if (cursorShape == Enum::CustomCursor)
+        view->setKeyboardCursorShape(Enum::CustomCursor);
 
     // cursor color
     if (profile->useCustomCursorColor()) {
