diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index bde2973..85b1798 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -128,7 +128,11 @@ set(konsole_LIBS ${KDE4_KPTY_LIBS}
                  ${KDE4_KIO_LIBS}
                  ${KDE4_KNOTIFYCONFIG_LIBS}
                  ${QT_QTXML_LIBRARY}
-                 ${X11_LIBRARIES})
+                 ${X11_LIBRARIES}
+		 qtlua)
+
+#Required for qtlua
+set(CMAKE_CXX_FLAGS "-fpermissive")
 
 if(HAVE_LIBKONQ)
   set(konsole_LIBS ${konsole_LIBS} ${LIBKONQ_LIBRARY})
diff --git a/src/EditProfileDialog.ui b/src/EditProfileDialog.ui
index 82298cf..92ea9cc 100644
--- a/src/EditProfileDialog.ui
+++ b/src/EditProfileDialog.ui
@@ -1109,6 +1109,16 @@
                 <string comment="The shape of the cursor">Underline</string>
                </property>
               </item>
+              <item>
+               <property name="text">
+                <string comment="The shape of the cursor">Custom Dynamic</string>
+               </property>
+              </item>
+              <item>
+               <property name="text">
+                <string comment="The shape of the cursor">Custom Static</string>
+               </property>
+              </item>
              </widget>
             </item>
             <item>
diff --git a/src/Enumeration.h b/src/Enumeration.h
index a9fab40..a39083f 100644
--- a/src/Enumeration.h
+++ b/src/Enumeration.h
@@ -85,7 +85,9 @@ public:
          */
         IBeamCursor     = 1,
         /** Draw a line underneath the cursor's position. */
-        UnderlineCursor = 2
+        UnderlineCursor = 2,
+        CustomCursorDynamic = 3,
+        CustomCursorStatic = 4
     };
 
     /** This enum describes the behavior of triple click action . */
diff --git a/src/TerminalDisplay.cpp b/src/TerminalDisplay.cpp
index 2235047..9d22017 100644
--- a/src/TerminalDisplay.cpp
+++ b/src/TerminalDisplay.cpp
@@ -72,6 +72,12 @@
 #include "SessionManager.h"
 #include "Session.h"
 
+// Custom cursor painting.
+#include <iterator>
+#include <QtLua/State>
+#include <QtLua/Value>
+#include "CustomCursor.h"
+
 using namespace Konsole;
 
 #ifndef loc
@@ -632,6 +638,7 @@ void TerminalDisplay::drawBackground(QPainter& painter, const QRect& rect, const
     painter.fillRect(scrollBarArea, _scrollBar->palette().background());
 }
 
+#include <iostream>
 void TerminalDisplay::drawCursor(QPainter& painter,
                                  const QRect& rect,
                                  const QColor& foregroundColor,
@@ -678,6 +685,36 @@ void TerminalDisplay::drawCursor(QPainter& painter,
                          cursorRect.top(),
                          cursorRect.left(),
                          cursorRect.bottom());
+
+    } else if (_cursorShape == Enum::CustomCursorDynamic) {
+
+	QFile file("test.lua");
+	QtLua::State state;
+	state.openlib(QtLua::AllLibs);
+
+	//state["painter"] = QtLua::Value(state, &painter, false);
+	state["painter"] = &painter;
+	state["cursorRect"] = &cursorRect;
+
+	/* Open the Lua script file. */
+	file.open(QIODevice::ReadOnly);
+
+	/* Execute the contents. */
+	try {
+		state.exec_chunk(file);
+	} catch (QtLua::String e) {
+
+		//attempt to index global 'cursorRect' (a number value)
+		std::cout << e.to_qstring().toUtf8().constData() << std::endl;
+	}
+	
+	/* Close the file.  */
+	file.close();
+    } else if (_cursorShape == Enum::CustomCursorStatic) {
+
+	/* This is defined in CustomCursor.h */
+    	CustomCursor::draw_cursor(painter, cursorRect);
+    
     }
 }
 
diff --git a/src/ViewManager.cpp b/src/ViewManager.cpp
index 0b43bb8..590cdf9 100644
--- a/src/ViewManager.cpp
+++ b/src/ViewManager.cpp
@@ -858,6 +858,10 @@ void ViewManager::applyProfileToView(TerminalDisplay* view , const Profile::Ptr
         view->setKeyboardCursorShape(Enum::IBeamCursor);
     else if (cursorShape == Enum::UnderlineCursor)
         view->setKeyboardCursorShape(Enum::UnderlineCursor);
+    else if (cursorShape == Enum::CustomCursorDynamic)
+        view->setKeyboardCursorShape(Enum::CustomCursorDynamic);
+    else if (cursorShape == Enum::CustomCursorStatic)
+        view->setKeyboardCursorShape(Enum::CustomCursorStatic);
 
     // cursor color
     if (profile->useCustomCursorColor()) {
